dialogue-core/src/main/java/com/palantir/dialogue/core/BlacklistingChannel.java

```diff
@@ -166,43 +136,59 @@ public String toString() {
 
     private final class BlacklistState {
         private final AtomicBoolean inProbation = new AtomicBoolean();
-        private final BlacklistUntil blacklistUntil;
+        private final long blacklistUntilNanos;
         private final Probation probation;
-        private final ScheduledFuture<?> future;
+        private final ScheduledFuture<?> scheduledFuture;
 
         BlacklistState(Duration duration, int probationPermits) {
-            this.blacklistUntil = new BlacklistUntil(ticker.read() + duration.toNanos());
+            this.blacklistUntilNanos = ticker.read() + duration.toNanos();
             this.probation = new Probation(probationPermits);
-            this.future = scheduler.schedule(this::maybeProgressAndGet, duration.toNanos(), TimeUnit.NANOSECONDS);
+            this.scheduledFuture =
+                    scheduler.schedule(this::maybeProgressAndGet, duration.toNanos(), TimeUnit.NANOSECONDS);
+        }
+
+        Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+            Optional<Probation> maybeProbation = maybeProgressAndGet();
+            if (!maybeProbation.isPresent()) {
+                return Optional.empty();
+            }
+            if (maybeProbation.get().acquireStartPermit()) {
+                log.debug("Probation channel request allowed");
+                return delegate.maybeExecute(endpoint, request)
+                        .map(future -> DialogueFutures.addDirectCallback(future, blacklistingCallback));
+            } else {
+                log.debug("Probation channel request not allowed");
+                return Optional.empty();
+            }
         }
 
-        BlacklistStage maybeProgressAndGet() {
+        Optional<Probation> maybeProgressAndGet() {
             if (inProbation.get()) {
-                return probation;
+                return Optional.of(probation);
             }
-            if (ticker.read() >= blacklistUntil.untilNanos) {
+            if (ticker.read() >= blacklistUntilNanos) {
                 if (inProbation.compareAndSet(false, true)) {
```

Comment:
elsewhere we have debug logs for interesting transitions - can we include one here for beginning probation?

Reply:
good call